src/Tasks/ResolveAssemblyReferences/Server/RarController.cs

```diff
@@ -0,0 +1,158 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.IO;
+using System.IO.Pipes;
+using System.Security.AccessControl;
+using System.Security.Principal;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Build.Framework;
+using Microsoft.Build.Shared;
+using Microsoft.Build.Tasks.ResolveAssemblyReferences.Contract;
+using Microsoft.Build.Tasks.ResolveAssemblyReferences.Services;
+using Microsoft.VisualStudio.Threading;
+
+using StreamJsonRpc;
+
+#nullable enable
+namespace Microsoft.Build.Tasks.ResolveAssemblyReferences.Server
+{
+    public sealed class RarController
+    {
+        private const int PipeBufferSize = 131072;
+
+        /// <summary>
+        /// Name of <see cref="NamedPipeServerStream"/>
+        /// </summary>
+        private readonly string _pipeName;
+
+        /// <summary>
+        /// Handler for all incoming tasks
+        /// </summary>
+        private readonly IResolveAssemblyReferenceTaskHandler _resolveAssemblyReferenceTaskHandler;
+
+        /// <summary>
+        /// Timeout for incoming connections
+        /// </summary>
+        private readonly TimeSpan Timeout = new TimeSpan(0, 15, 0);
```

Comment:
nit: `TimeSpan.FromMinutes(15)` would be more readable.

Reply:
Updated